### PR TITLE
add endpoint

### PR DESCRIPTION
**Pull request by pingfengluo**
Add endpoint as https://github.com/k2-fsa/sherpa/issues/229

### CODE REVIEW

**Comment by pingfengluo** [comment]
Add endpoint as #229

File: sherpa/cpp_api/endpoint.cc [hunk 1]
```diff
@@ -38,6 +39,32 @@ static bool RuleActivated(const EndpointRule& rule,
   return ans;
 }
 
+static void RegisterEndpointRule(
+    ParseOptions *po, EndpointRule *rule, const std::string rule_name) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/endpoint.cc:43]:
```suggestion
    ParseOptions *po, EndpointRule *rule, const std::string &rule_name) {
```

File: sherpa/cpp_api/online-recognizer.h [hunk 2]
```diff
@@ -46,6 +46,8 @@ struct OnlineRecognizerConfig {
   /// GPU for computation
   bool use_gpu = false;
 
+  bool use_endpoint = false;
```
Comment (csukuangfj) [inline sherpa/cpp_api/online-recognizer.h:49]: Could you also make it configurable as a command line argument?

File: sherpa/csrc/online-stream.cc [hunk 3]
```diff
@@ -80,6 +80,12 @@ class OnlineStream::OnlineStreamImpl {
 
   int32_t &GetNumTrailingBlankFrames() { return num_trailing_blank_frames_; }
 
+  const int32_t GetFrameShift() { return frame_shift_ms_; }
```
Comment (csukuangfj) [inline sherpa/csrc/online-stream.cc:83]:
```suggestion
 int32_t GetFrameShift() const { return frame_shift_ms_; }
```

File: sherpa/cpp_api/online-stream.h [hunk 4]
```diff
@@ -149,6 +149,12 @@ class OnlineStream {
   // The returned reference is valid as long as this object is alive.
   int32_t &GetNumTrailingBlankFrames();
 
+  const int32_t GetFrameShift();
```
Comment (csukuangfj) [inline sherpa/cpp_api/online-stream.h:153]:
```suggestion
  int32_t GetFrameShift() const;
```

Please also describe whether the returned value is in seconds or in milliseconds.

File: sherpa/cpp_api/online-recognizer.cc [hunk 5]
```diff
@@ -340,12 +337,11 @@ class OnlineRecognizer::OnlineRecognizerImpl {
...
-  bool IsEndpoint(const int num_frames_decoded,
-      const int trailing_silence_frames,
-      const float frame_shift_in_seconds) const {
-    return endpoint_->IsEndpoint(num_frames_decoded,
-                                 trailing_silence_frames,
-                                 frame_shift_in_seconds);
+  bool IsEndpoint(OnlineStream * s) const {
+    return endpoint_->IsEndpoint(s->GetNumProcessedFrames()
+                                 - s->GetStartFrame(),
+                                 s->GetNumTrailingBlankFrames() * 4,
```
Comment (csukuangfj) [inline sherpa/cpp_api/online-recognizer.cc:343]:
```suggestion
                                 s->GetNumTrailingBlankFrames() *  model_->SubsamplingFactor(),
```

File: sherpa/csrc/online-stream.cc [hunk 6]
```diff
@@ -90,7 +96,12 @@ class OnlineStream::OnlineStreamImpl {
   torch::Tensor decoder_out_;
   int32_t num_processed_frames_ = 0;       // before subsampling
   int32_t num_trailing_blank_frames_ = 0;  // after subsampling
-  // int32_t frame_shift_ms_ = 10;            // before subsampling
+  int32_t frame_shift_ms_ = 10;            // before subsampling
```
Comment (csukuangfj) [inline sherpa/csrc/online-stream.cc:99]: Shall we use the value from `fbank_->FrameOpts()`.?

File: .github/scripts/run-online-transducer.sh [hunk 7]
```diff
@@ -204,6 +211,17 @@ time ./build/bin/sherpa-online \
...
 log "=========================================================================="
+
+# For Endpoint
+for m in greedy_search modified_beam_search fast_beam_search; do
+  time ./build/bin/sherpa-online \
+    --decoding-method=$m \
+    --nn-model=$repo/exp/cpu_jit.pt \
+    --tokens=$repo/data/lang_char/tokens.txt \
+    --use-endpoint=true \
+    $long_audio_repo/10-minutes.wav
```
Comment (csukuangfj) [inline .github/scripts/run-online-transducer.sh:224]: 10-minutes.wav is for English, while this model is for Chinese. I am afraid it is not able to recognize the wave.

**Comment by csukuangfj** [comment]
Thanks!

**Pull request merged**